### PR TITLE
Rewrote part of GA4 to support add/remove quantity from cart and cart_updates

### PR DESCRIPTION
**Pull request by fballiano**
The previous implementation:
- didn't support the add_to_cart quantity parameter, this was a problem when there are products with minimum_quantity (meaning you will have to add X amount to the cart, not only 1, while ga4 was tracking only 1).
- was forcing a `Mage::getModel('catalog/product')->load()` for every product, which is now removed since it was unneeded
- wasn't working with cart_updates (when you change the quantity of a product in the cart page), now this PR calculates the difference and triggers add_to_cart or remove_to_cart events

This version should be much more complete than before, hope you like it.

### CODE REVIEW

**Comment by fballiano** [comment]
@empiricompany something for you :-D

**Comment by empiricompany** [comment]
Interesting implementation! As soon as I have a minute, I will test it.

**Comment by empiricompany** [comment]
I still need to test it but i love how you avoided loading the product and controls in quantities...

even though for me there is too much duplicated code.
In the observer, you compile an array, and in the block, you call individual items and format it.

What do you think about compiling a session variable that contains arbitrary datalayer data including also event type remove_from_cart / add_to_cart? 
For example:
Mage::getSingleton('core/session')->getAdditionalDataLayerData();

It will contain something like this:

```
$getAdditionalDataLayerData = [
    'event' => [
        'add_to_cart' => [
            'item_id' => $product->getId(),
            'sku' => $product->getSku(),
            'name' => $product->getName(),
            'qty' => (int) $addedQty,
            'price' => Mage::helper('googleanalytics')->formatPrice($product->getFinalPrice()), // new helper to format price
            'manufacturer' => $product->getAttributeText('manufacturer') ?: '',
            'category' => Mage::helper('googleanalytics')->getLastCategoryName($product)
        ],
        'remove_from_cart ' => [
            'item_id' => $product->getId(),
            'sku' => $product->getSku(),
            'name' => $product->getName(),
            'qty' => (int) $removedQty,
            'price' => Mage::helper('googleanalytics')->formatPrice($product->getFinalPrice()), // new helper to format price
            'manufacturer' => $product->getAttributeText('manufacturer') ?: '',
            'category' => Mage::helper('googleanalytics')->getLastCategoryName($product)
        ],
         'whatever_youwant ' => [ 
           //data from 3rd party 
            ..
    ]
];
```
PS: i also included an helper to format the price and cast to int for qty

then catch it in the block / ga and call gtag(... for every items?

It would be more flexible and could be used in other parts of the code and third-party modules as well.

**Comment by fballiano** [comment]
I've removed 2 methods from the observer that were actually duplicate (apart from having "added" or "removed" as variable names) and you tell me it's too much duplicate code ehheheheh

At a first glance I woulnd't have code that "formats" the data in an observer, that's why I kept all of that in the block.

I'll think it out.

Anythis this PR still removes a load in the loop, fixes the quantity bug and adds the updates_cart functionality..

**Comment by empiricompany** [comment]
> I've removed 2 methods from the observer that were actually duplicate (apart from having "added" or "removed" as variable names) and you tell me it's too much duplicate code ehheheheh
> 
> At a first glance I woulnd't have code that "formats" the data in an observer, that's why I kept all of that in the block.
> 
> I'll think about it.
> 
> Anythis this PR still removes a load in the loop, fixes the quantity bug and adds the updates_cart functionality..

Ok, I will present my point of view in a separate PR.

File: app/code/core/Mage/GoogleAnalytics/Block/Ga.php [hunk 1]
```diff
@@ -137,61 +137,50 @@ protected function _getEnhancedEcommerceDataForAnalytics4()
...
+                $eventData['value'] = number_format($removedProduct['price'] * $removedProduct['qty'], 2, '.', '');
                 $eventData['items'] = [];
                 $_item = [
-                    'item_id' => $_removedProduct->getSku(),
-                    'item_name' => $_removedProduct->getName(),
-                    'price' => number_format($_removedProduct->getFinalPrice(), 2, '.', ''),
+                    'item_id' => $removedProduct['sku'],
+                    'item_name' => $removedProduct['name'],
+                    'price' => number_format($removedProduct['price'], 2, '.', ''),
+                    'quantity' => $removedProduct['qty'],
```
Comment (empiricompany) [inline app/code/core/Mage/GoogleAnalytics/Block/Ga.php:151]: add specific cast to int

`'quantity' => (int) $removedProduct['qty'],`

File: app/code/core/Mage/GoogleAnalytics/Block/Ga.php [hunk 2]
```diff
@@ -137,61 +137,50 @@ protected function _getEnhancedEcommerceDataForAnalytics4()
...
+                $eventData['value'] = number_format($_addedProduct['price'] * $_addedProduct['qty'], 2, '.', '');
                 $eventData['items'] = [];
                 $_item = [
-                    'item_id' => $_addedProduct->getSku(),
-                    'item_name' => $_addedProduct->getName(),
-                    'price' => number_format($_addedProduct->getFinalPrice(), 2, '.', ''),
+                    'item_id' => $_addedProduct['sku'],
+                    'item_name' => $_addedProduct['name'],
+                    'price' => number_format($_addedProduct['price'], 2, '.', ''),
+                    'quantity' => $_addedProduct['qty'],
```
Comment (empiricompany) [inline app/code/core/Mage/GoogleAnalytics/Block/Ga.php:177]: `'quantity' => (int) $_addedProduct['qty'],`

**Review by empiricompany — CHANGES_REQUESTED** [review]

**Comment by fballiano** [comment]
the formatprice helper is for sure a good idea, but I'd still keep the formatting in the block, and not expect data already formatted (if we give access to 3rd party code to inject data, they could forget to format the data)

File: app/code/core/Mage/GoogleAnalytics/Block/Ga.php (hunk 1 above)
Comment (fballiano) [reply app/code/core/Mage/GoogleAnalytics/Block/Ga.php]: done

**Comment by fballiano** [comment]
I've added the `formatPrice` helper method.

**Comment by empiricompany** [comment]
> I've added the `formatPrice` helper method.

needs a space 
`'quantity' => (int) $_addedProduct['qty'],`

**Comment by fballiano** [comment]
> > I've added the `formatPrice` helper method.
> 
> needs a space `'quantity' => (int) $_addedProduct['qty'],`

it was already fixed before your comment ;-)

**Comment by empiricompany** [comment]
> > > I've added the `formatPrice` helper method.
> > 
> > 
> > needs a space `'quantity' => (int) $_addedProduct['qty'],`
> 
> it was already fixed before your comment ;-)

great! comment was hidden

**Comment by empiricompany** [comment]
first test:
i have added one product with qty = 2

datalayer is ok:

```
(
    [0] => gtag('event', 'add_to_cart', {"currency":"USD","value":"10.00","items":[{"item_id":"hbm001","item_name":"Around the World in 80 Days","price":"5.00","quantity":2,"item_brand":"","item_category":"Books & Music"}]});
    [1] => gtag('event', 'view_item', {"currency":"USD","value":"5.00","items":[{"item_id":"hbm001","item_name":"Around the World in 80 Days","list_name":"Product Detail Page","item_category":"Books & Music","price":"5.00"}]});
)
```

then i have updated qty from 2 to 4
but datalayer is this:

```
(
    [0] => gtag('event', 'add_to_cart', {"currency":"USD","value":"10.00","items":[{"item_id":"hbm001","item_name":"Around the World in 80 Days","price":"5.00","quantity":2,"item_brand":"","item_category":"Books & Music"}]});
    [1] => gtag('event', 'add_to_cart', {"currency":"USD","value":"10.00","items":[{"item_id":"hbm001","item_name":"Around the World in 80 Days","price":"5.00","quantity":2,"item_brand":"","item_category":"Books & Music"}]});
    [2] => gtag('event', 'view_item', {"currency":"USD","value":"5.00","items":[{"item_id":"hbm001","item_name":"Around the World in 80 Days","list_name":"Product Detail Page","item_category":"Books & Music","price":"5.00"}]});
)
```

I'm not sure how the datalayer should be filled in this case

**Comment by fballiano** [comment]
there should be 1 add_to_cart with 2qty, because you're adding 2 more products to the cart, so it seems there's a duplicate event, weird, i'll check it out (i'm sure I tested that yesterday, before today's mods)

**Comment by empiricompany** [comment]
if i retouch the qty from 4 to 2
datalayer seems wrong:

```
(
    [0] => gtag('event', 'remove_from_cart', {"currency":"USD","value":"10.00","items":[{"item_id":"hbm001","item_name":"Around the World in 80 Days","price":"5.00","quantity":2,"item_brand":"","item_category":"Books & Music"}]});
    [1] => gtag('event', 'remove_from_cart', {"currency":"USD","value":"10.00","items":[{"item_id":"hbm001","item_name":"Around the World in 80 Days","price":"5.00","quantity":2,"item_brand":"","item_category":"Books & Music"}]});
    [2] => gtag('event', 'view_item', {"currency":"USD","value":"5.00","items":[{"item_id":"hbm001","item_name":"Around the World in 80 Days","list_name":"Product Detail Page","item_category":"Books & Music","price":"5.00"}]});
)
```

**Comment by empiricompany** [comment]
> 

i have modified the qty from rwd theme with ajax in cart popup, then i reloaded the page to catch and send new datalayer events

**Comment by empiricompany** [comment]
> there should be 1 add_to_cart with 2qty, because you're adding 2 more products to the cart, so it seems there's a duplicate event, weird, i'll check it out (i'm sure I tested that yesterday, before today's mods)


 i have modified the qty from rwd theme with ajax in cart popup, then i reloaded the page to catch and send new datalayer events

**Comment by fballiano** [comment]
I re-tested right now and I don't see your problem, ex modifying qty from 3 to 5:

Screenshot 2023-07-12 alle 14 15 03

are you on rwd?

**Comment by fballiano** [comment]
ah ok, in the minicart, I'll test it out

**Comment by empiricompany** [comment]
> I re-tested right now and I don't see your problem, ex modifying qty from 3 to 5:
> Screenshot 2023-07-12 alle 14 15 03
> 
> are you on rwd?

yes on rwd but seems you are on cart page 
i've tested whitin cart popup

**Comment by empiricompany** [comment]
> > I re-tested right now and I don't see your problem, ex modifying qty from 3 to 5:
> > Screenshot 2023-07-12 alle 14 15 03
> > are you on rwd?
> 
> yes on rwd but seems you are on cart page i've tested whitin cart popup

i confirm in the cart page work ok

i've expected this problem in previous implementation for this was added array_unique cause event is fired twice i don't know the reason

`$_addedProducts = array_unique($_addedProducts);`

**Comment by fballiano** [comment]
I've seen that but the array unique could create problems when adding multiple of the same product programmatically and I wanted to avoid that.

but it seems here something is running twice.

**Comment by fballiano** [comment]
there's a double save of the quote item in the ajax cart controller, i'll fix it in another pr, but i'll add some sort of array_unique here too

**Comment by empiricompany** [comment]
i report also that in view_cart event total value is not multiplied by qty

here we have 10 qty of item but the total value of the cart is 5,  the correct value must be 50 (after currency)
```
(
    [0] => gtag('event', 'view_cart', {"currency":"USD","value":"5.00","items":[{"item_id":"hbm001","item_name":"Around the World in 80 Days","price":"5.00","quantity":10,"item_category":"Books & Music"}]});
)
```

**Comment by fballiano** [comment]
> i report also that in view_cart event total value is not multiplied by qty
> 
> here we have 10 qty of item but the total value of the cart is 5, the correct value must be 50 (after currency)
> 
> ```
> (
>     [0] => gtag('event', 'view_cart', {"currency":"USD","value":"5.00","items":[{"item_id":"hbm001","item_name":"Around the World in 80 Days","price":"5.00","quantity":10,"item_category":"Books & Music"}]});
> )
> ```

it wasn't touched by this PR, so, if this is a bug, we'll check it in a later one.

**Comment by empiricompany** [comment]
> > i report also that in view_cart event total value is not multiplied by qty
> > here we have 10 qty of item but the total value of the cart is 5, the correct value must be 50 (after currency)
> > ```
> > (
> >     [0] => gtag('event', 'view_cart', {"currency":"USD","value":"5.00","items":[{"item_id":"hbm001","item_name":"Around the World in 80 Days","price":"5.00","quantity":10,"item_category":"Books & Music"}]});
> > )
> > ```
> 
> it wasn't touched by this PR, so, if this is a bug, we'll check it in a later one.

yes is a bug from previous code, simply mod on line 289:

`$eventData['value'] += $_product->getFinalPrice() * $productInCart->getQty();`

File: app/code/core/Mage/GoogleAnalytics/Block/Ga.php [hunk 3]
```diff
@@ -284,21 +274,21 @@ protected function _getEnhancedEcommerceDataForAnalytics4()
...
                     $_item['item_brand'] = $_product->getAttributeText('manufacturer');
                 }
 
-                $itemCategory = Mage::helper('googleanalytics')->getLastCategoryName($_product);
+                $itemCategory = $helper->getLastCategoryName($_product);
                 if ($itemCategory) {
                     $_item['item_category'] = $itemCategory;
                 }
                 array_push($eventData['items'], $_item);
                 $eventData['value'] += $_product->getFinalPrice();
```
Comment (empiricompany) [inline app/code/core/Mage/GoogleAnalytics/Block/Ga.php:289]: `$eventData['value'] += $_product->getFinalPrice() * $productInCart->getQty();`

Comment (fballiano) [reply app/code/core/Mage/GoogleAnalytics/Block/Ga.php]: already done

**Comment by fballiano** [comment]
I've just committed a sort of array_unique, using the Mage::registry and arrayobject (registry doesn't allow to overwrite data, arrayobjects is by-reference).

didn't use a array_unique because the previous implementation was checkin it "after" all the computations, while I wanted to exit the method as soon as possible

**Review by roby255 — APPROVED** [review]

**Comment by fballiano** [comment]
I got some tests from a customer that is in the need for this and this latest version seems to work ok with all use cases.

**Comment by empiricompany** [comment]
> I got some tests from a customer that is in the need for this and this latest version seems to work ok with all use cases.

I have tested the latest changes, and it works correctly in various scenarios.
PS: It's amazing how it calculates the quantity differences and adds or removes items from the cart accordingly!
Great work @fballiano !

**Review by empiricompany — APPROVED** [review]

**Review by ADDISON74 — APPROVED** [review]

**Comment by fballiano** [comment]
merging per 1green+2gray :-)

**Pull request merged**